### PR TITLE
[Inductor][Bucketing] Isolate bucketing traces from ambient unbacked symbols

### LINKED ISSUE #183679: [Inductor][Bucketing] Nested bucket traces should ignore ambient pending unbacked symbols
## Summary

Inductor collective bucketing can fail when its nested `make_fx` trace observes pending unbacked symbols from an outer fake/dynamic-shape trace.

Bucketing traces small collective rewrite helpers while reusing the surrounding `FakeTensorMode`. If that shared `ShapeEnv` already has pending fresh unbacked symbols from an unrelated outer trace, `compute_unbacked_bindings` can require those ambient symbols to be bound by the bucket trace outputs. The bucket helper itself may only operate on static tensors, so those symbols are unrelated and cannot be bound by the nested trace.

## Reproducer

```python
import torch
from torch._C import FileCheck
from torch._inductor.fx_passes.bucketing import _trace as bucketing_trace
from torch._subclasses import FakeTensorMode
from torch.fx.experimental.symbolic_shapes import ShapeEnv


fake_mode = FakeTensorMode(allow_non_fake_inputs=True, shape_env=ShapeEnv())
x = fake_mode.from_tensor(torch.randn(2, 2), static_shapes=True)

ambient = fake_mode.shape_env.create_unbacked_symint().node.expr
assert ambient in fake_mode.shape_env.pending_fresh_unbacked_symbols

gm = bucketing_trace(lambda x: x + 1, (x,))

assert ambient in fake_mode.shape_env.pending_fresh_unbacked_symbols
FileCheck().check("aten.add").run(gm.code)
print(gm.code)
```

## Expected behavior

The nested bucketing trace should only be responsible for symbols created by the bucketing trace itself. Ambient pending unbacked symbols from the surrounding trace should remain pending for the outer trace and should not be required in the bucket helper outputs.

## Actual behavior

The nested `make_fx` trace can fail with `PendingUnbackedSymbolNotFound` for an ambient unbacked symbol that is unrelated to the bucket inputs and outputs.

## Notes

The bucketing trace should snapshot and clear pending/ignorable fresh unbacked symbols around the nested `make_fx` call, then restore the ambient state afterward.

cc @chauhang @penguinwu @eellison @aorenste @ezyang @bobrenjc93 @aditvenk @laithsakka @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @aakhundov @coconutruben @jataylo @bdhirsh

### PR DESCRIPTION
**Pull request by sanketpurandare**
Stacked PRs:
 * #183545
 * #183544
 * __->__#183495

--- --- ---

### [Inductor][Bucketing] Isolate bucketing traces from ambient unbacked symbols

Fixes https://github.com/pytorch/pytorch/issues/183679

Bucketing traces small collective rewrite helpers with make_fx while reusing the FakeTensorMode from the surrounding graph. When that shared ShapeEnv already has pending fresh unbacked symbols from an outer dynamic-shape trace, make_fx's unbacked binding validation can try to account for symbols that are unrelated to the bucket inputs and outputs.

This shows up when a static bucket trace runs after a graph trace that introduced unbacked activation symbols: the bucket trace itself only sees static parameter or gradient tensors, but compute_unbacked_bindings observes the ShapeEnv's global pending_fresh_unbacked_symbols and errors on ambient symbols it cannot bind.

Snapshot and clear pending/ignorable fresh unbacked symbols around the nested bucketing make_fx trace, then restore the ambient state afterward. This keeps bucketing responsible only for symbols produced by its own trace while preserving the outer ShapeEnv bookkeeping for the caller.

Add a regression that creates an ambient pending unbacked symbol in a FakeTensorMode, traces a static bucketing helper, and verifies the ambient symbol remains pending after the trace.

Test Plan:
python test/distributed/test_inductor_collectives.py -k test_trace_ignores_ambient_pending_unbacked_symbols -v
lintrunner --config=.lintrunner.toml torch/_inductor/fx_passes/bucketing.py test/distributed/test_inductor_collectives.py

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Review by IvanKobzarev — APPROVED** [review]
lgtm
